blueprints/varnish-support.md

```diff
@@ -0,0 +1,117 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+# Add Varnish Cache Support
+
+## Problem Description
+
+Currently Traffic Control uses Traffic Server as the underlying cache server. We can expand that by introducing Varnish cache as an option for the cache server used with its great performance, robustness and modularity.
+
+## Proposed Change
+
+From a high level point of view, ATS operates based on configuration files that describe in details how it should work and interact with other servers in the cache hierarchy. These configuration files are managed and generated using `t3c` components that utilize Traffic Ops APIs to get profiles and parameters data required for the configuration files. The proposed change is to use the same data fetched from Traffic Ops APIs to generate configuration files for Varnish cache with almost the same functionality.
+
+Note: the changes should not affect existing components but rather build on them.
+
+### Traffic Portal Impact
+
+n/a
+
+### Traffic Ops Impact
+
+n/a
+
+#### REST API Impact
+
+n/a
+
+#### Client Impact
+
+n/a
+
+#### Data Model / Database Impact
+
+- A new profile type will be added `VARNISH_PROFILE` but it won't affect the model or other structures in `lib/go-tc`
+- `DeliveryService` structs contain fields related to ATS like `remapText`. It will be parsed and translated to Varnish configuration.
+
+### ORT Impact
+
+- `go-varnish` package will be developed to handle generating configuration files for Varnish, Hitch, `varnishncsa` and `logrotate`. For detailed description of mapping configuration files from ATS to Varnish refer to [Varnish Support](https://github.com/apache/trafficcontrol/wiki/Varnish-Support) wiki.
+- New subcommands will be added to `t3c` including `t3c-varnish-generate` that works similar to `t3c-generate` but instead will depend on `go-varnish` package to generate different configuration files and `t3c-varnish-apply` that handles applying these configuration files, restarting and reloading different services running.
```

Comment:
why not just let `t3c-generate` do that? Add, like `--cache=varnish` if it can't auto-detect it. Although, maybe that's what the profile type could be used for? Tells `t3c` it's a varnish cache?

Reply:
We could definitely reuse `t3c-generate` and `t3c-apply`, however flags used with these commands are made for ATS and I assume new Varnish stuff will introduce new flags too. So, it might be confusing to have some flags work with `--cache=ats` and some with `--cache=varnish` and possibly new cache servers added to ATC. Wouldn't it be better if `t3c` decides from profile type if it uses `t3c-varnish-apply` or basically for any cache server `t3c-*-apply` instead of having all logic in the same command?

Reply:
anything in t3c generate that currently works a specific way because it only supports ATS should be rewritten/reworded to also support varnish, if we're gonna support varnish. Just as an example, there's an option `--dir` that says it's used to specify the ATS configuration directory - it should instead just say that it's used to specify the cache server configuration directory.

Reply:
Well, for options used in all cache servers it is no problem to just rewrite it. For options that is not used in Varnish (or other cache servers in the future) like `disable-parent-config-comments` or `ats-version` it won't make sense in other cache servers context. The same case also will happen when other options are introduced in Varnish case.
I guess we could restructure the documentation and help option to show options per cache server type:
```
## common options
-dir
-silent
...
## ats options
-ats-version
...
## varnish options
...
```
What do you think?

Reply:
`--ats-version` should probably be just `--cache-version` or something, since - and I don't know anything about Varnish - that probably still has consequences no matter which software you use.

As for `disable-parent-config-comments`, I don't really know what the purpose of that is. What comments does it include or not include? If it's just a comment, why would you ever care enough to even use that option?

In any case, there's probably some example that very clearly and obviously only has meaning to an ATS cache - in that case, I'd vote we just note that does nothing for Varnish caches. Or, rather, only does things for ATS caches. I'm not too concerned with how the docs get structured as long they exist, are accurate, and make sense.